py-data-manipulation.Rmd

```diff
@@ -0,0 +1,457 @@
+
+1. **Understanding data frame structure**
+
+Based on the output of `dog_licenses.info()`, can you answer the following questions?
+
+ - What is the class of the object `dog_licenses` (where
+   `dog_licenses = pd.read_csv("data/nyc-dog-licensing.csv")`)?
+ - How many rows and how many columns are in this object?
+ - Why is there not the same number of rows (observations) for each column?
+
+2. **Finding messy values**
+
+Get a list of the unique values in the `breed_name` column of the dog license
+data using `value_counts()` or `unique()`. Sort this list alphabetically. (Hint:  
+use the `sort_values()` function.) Look at the first 50 values. Do you see any
+breed names that should be combined?
+
+2. **Subsetting data by row**
+
+- Extract the 200th and 201st row of the dog license dataset and assign the
+resulting data frame to a new variable name (i.e. `data_200_201`). Remember
+that Python indexing starts at 0!
+
+- With the data frame `dog_licenses = pd.read_csv("data/nyc-dog-licensing.csv")`:
+    - How can you get the same result as from `dog_licenses.head()` by using row slices
+    instead of the `head()` method?
+
+- There are at least three distinct ways to extract the last row of a data
+frame. How many can you come up with?
+
+3. **Selecting and filtering data**
+
+- Subset the dog license data frame `data` to include only female dogs and retain
+only the columns 'animal_name', 'animal_gender' and 'breed_name'. There are
+multiple ways this could be done.
+
+- How many dog licenses belong to dogs named "Queen" that live in "Queens"?
+
+3. **Error messages**
+
+Run the following code. What type of error does it generate?
+
+```python
+dog_licenses.loc[0:10, 'animal_name', 'breed_name']
+```
+
+Edit the code to prevent it from causing an error.
+
+4. **Creating new columns**
+
+Create a new data frame from the dog license data frame `dog_licenses` that contains only the `animal_birth_month`, `animal_name`, and `breed_name` columns.
+Add two new columns to this new data frame: one called `birth_year` which contains just the year from the `animal_birth_month` column, and one
+called `birth_month` which contains just the month from the `animal_birth_month` column.
+
+5. **groupby()**
+
+  1. Use `groupby()` and `agg()` with the dog license data frame `dog_licenses` after adding the `age_when_licensed` column to find the mean, min, and max age when a dog was licensed in days grouped by `borough`. Hint: the pandas `timedelta` format of the `age_when_licensed` column will make calculating numeric quantities difficult; try creating a new column for the licensed age in days using `dog_licenses['age_when_licensed'].dt.days`.
+
+  Which borough has the largest mean time until licensing? Return the columns `breed_name`, `borough`, and `age_when_licensed`. Hint: Look into the `idxmax()` method.
+
+  2. How many dogs are female and how many are male?
+
+
+## Key Points {#py-data-manipulation-keypoints}
```

Comment:
```suggestion
## Key Points {#py-data-manipulation-keypoints}

```{r, child="keypoints/py-data-manipulation.md"}
```

Need to end with ```, the suggestion won't allow me to add it.